### PR TITLE
fix: use deterministic keys for MCP server tool routing

### PR DESCRIPTION
**Pull request by KeWang0622**
### Related Issue

**Issue:** #8087

### Description

MCP tool function names embed a server key (e.g. `cg9IjN0mcp0health_check`). Today this key is generated with `nanoid(5)` — random and ephemeral. After an extension restart or MCP reconnect the key changes, but the model may still have old tool names cached in its context. This causes "No connection found for server" routing failures.

This PR replaces the random nanoid with a **deterministic djb2 hash** of the server name. The same server always gets the same key, so cached tool names remain valid across restarts.

**Changes:**
- Extract `hashServerName()` / `buildServerKey()` into a standalone `server-key.ts` utility (zero dependencies, easy to test)
- Replace `nanoid(5)` call in `McpHub.getMcpServerKey()` with `buildServerKey()`
- Remove unused `nanoid` import from `McpHub.ts`
- Add 10 unit tests covering determinism, output format, edge cases

**Key design decisions:**
- **djb2 hash** — fast, good distribution for short strings, well-established
- **base-36 output** — only `[a-z0-9]`, safe for function-name identifiers across all providers
- **`c` prefix preserved** — Gemini requires function names to start with a letter
- **Same 6-char length** (`c` + 5 chars) — no change to tool name format or length budgets
- **In-memory Map retained** — still used for O(1) reverse lookup in `getMcpServerByKey()`

### Test Procedure

1. Configure 2+ MCP servers with native tool calls enabled
2. Start Cline, list tools — observe stable tool name keys
3. Reload window / restart extension host
4. List tools again — keys should be identical to step 2
5. Attempt MCP tool call — should route correctly (previously failed)

**Unit tests:**
```bash
TS_NODE_PROJECT=./tsconfig.unit-test.json npx mocha \
  --require ts-node/register --require source-map-support/register \
  --spec "src/services/mcp/__tests__/McpHub.serverKeys.test.ts" --no-config
```
All 10 passing.

### Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)

### Pre-flight Checklist

- [x] Changes are limited to a single feature/bugfix
- [x] Tests are passing
- [x] Code is formatted and linted (biome)
- [x] I have reviewed [contributor guidelines](https://github.com/cline/cline/blob/main/CONTRIBUTING.md)

Co-Authored-By: Claude Opus 4.6 (1M context) <noreply@anthropic.com>

### CODE REVIEW

File: src/services/mcp/McpHub.ts [hunk 1]
```diff
@@ -134,14 +137,19 @@ export class McpHub {
...
+		const uid = buildServerKey(server)
 		McpHub.mcpServerKeys.set(uid, server)
 		return uid
 	}
 
+	/**
+	 * Expose hashServerName for backward compatibility and direct access.
+	 * @see {@link hashServerName} in ./server-key.ts
+	 */
+	static hashServerName = hashServerName
```
[the remark this answers is not in the document]
Comment (KeWang0622) [reply src/services/mcp/McpHub.ts]: Fixed in b0c93053 — removed the unused `McpHub.hashServerName` static property and its `hashServerName` import. The function remains exported from `server-key.ts` where the tests import it directly.

File: src/services/mcp/server-key.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ * base-36 output (0-9a-z) is safe for function-name identifiers.
+ */
+export function hashServerName(name: string): string {
+	let hash = 5381
+	for (let i = 0; i < name.length; i++) {
+		hash = (hash * 33) ^ name.charCodeAt(i)
+	}
+	// Convert to unsigned 32-bit, then to base-36 (0-9a-z), take first 5 chars
+	return (hash >>> 0).toString(36).padStart(5, "0").slice(0, 5)
+}
```
[the remark this answers is not in the document]
Comment (KeWang0622) [reply src/services/mcp/server-key.ts:27]: Fixed in b0c93053 — added collision detection before `mcpServerKeys.set()` in `getMcpServerKey`. When two different server names hash to the same key, it now emits a `Logger.warn` identifying both server names and the colliding key, before overwriting.

**Comment by Kirtofu** [comment]
Hi @KeWang0622 — status update on this PR, verified against current `main` (`8bbdde2`, 2026-08-14).

**The encode/decode path this PR patches has been removed upstream, so this change is superseded by #12773** ("Remove vestigial MCP server-key machinery from McpHub", merged 2026-07-31, one day after the maintainer bot's last comment here).

Evidence on `main`:

- `src/services/mcp/McpHub.ts` no longer exists (the whole legacy `src/` tree was removed by the extension-host restructure) — which is also why this PR reports CONFLICTING. All three changed files live under the removed tree.
- `mcpServerKeys`, `getMcpServerKey` / `getMcpServerByKey`, and `c${nanoid(5)}` have **zero references anywhere in the repository** (checked all files, not just `.ts`, across `apps/` and `sdk/`).
- MCP tool names now come from `defaultMcpToolNameTransform` in `sdk/packages/core/src/extensions/mcp/name-transform.ts`: `${serverName}__${toolName}`, with a sha1-8 suffix only when sanitization or the 64-char cap requires it.
- Tool execution closes over the server name directly (`sdk/packages/core/src/extensions/mcp/tools.ts` → `provider.callTool({ serverName, toolName })`), and the VSCode layer routes `callTool(serverName, toolName, ...)` by name (`apps/vscode/src/services/mcp/McpHub.ts`). There is nothing left to encode/decode, so nothing goes stale across restarts.

That means #8087 is fixed on main by architecture — #12773's description states exactly that ("supersedes #10254").

The diagnosis and tests here were correct for the code as it existed when this was opened, and the collision point raised in review (overwriting the registry on a hash collision turns a session-scoped miss into a permanent one) is a genuinely good catch that also matches what the maintainer bot requested. But there is no longer a place in the code to apply a "better hash" — the right change was deletion, and it has been done.

Recommendation: close this PR as superseded by #12773. If you still see routing instability under the new `serverName__toolName` scheme, a regression test around reconnect + `defaultMcpToolNameTransform` would be the right shape for a follow-up.

Also FYI: issue #8087 is still open even though #12773 fixed the root cause — a maintainer will probably want to close it.
